contents/quantum_systems/code/c++/energy.cpp

```diff
@@ -1,50 +1,54 @@
+#include <complex>
 #include <vector>
 #include <complex>
+#include <cstring>
 
 #include <fftw3.h>
 
 void fft(std::vector<std::complex<double>> x, bool inverse) {
-    std::vector<std::complex<double>> y(x.size());
-    for (size_t i = 0; i < x.size(); i++) {
-        y.push_back(std::complex(0.0, 0.0));
-    }
-    
+    std::vector<std::complex<double>> y(x.size(), std::complex<double>(0.0, 0.0));
+
     fftw_plan p;
 
-    p = fftw_plan_dft_1d(x.size(), reinterpret_cast<fftw_complex*>(x.data()),
-                         reinterpret_cast<fftw_complex*>(y.data()),
-                         (inverse ? FFTW_BACKWARD : FFTW_FORWARD), FFTW_ESTIMATE);
+    fftw_complex *in = reinterpret_cast<fftw_complex*>(x.data());
+    fftw_complex *out = reinterpret_cast<fftw_complex*>(y.data());
+
+    p = fftw_plan_dft_1d(x.size(), in, out,
+                        (inverse ? FFTW_BACKWARD : FFTW_FORWARD), FFTW_ESTIMATE);
+
 
     fftw_execute(p);
     fftw_destroy_plan(p);
 
     for (size_t i = 0; i < x.size(); ++i) {
-        x[i] = y[i] / sqrt((double)x.size());
+        x[i] = y[i] / sqrt(static_cast<double>(x.size()));
     }
 }
 
-double calculate_energy(std::vector<std::complex<double>> wfc, std::vector<std::complex<double>> h_r,
-                        std::vector<std::complex<double>> h_k, double dx, size_t size) {
+double calculate_energy(std::vector<std::complex<double>> wfc,
+                        std::vector<std::complex<double>> h_r,
+                        std::vector<std::complex<double>> h_k,
+                        double dx, size_t size) {
     std::vector<std::complex<double>> wfc_k(wfc);
     std::vector<std::complex<double>> wfc_c(size);
     fft(wfc_k, false);
 
     for (size_t i = 0; i < size; ++i) {
-        wfc_c.push_back(conj(wfc[i]));
+        wfc_c[i] = conj(wfc[i]);
     }
 
     std::vector<std::complex<double>> energy_k(size);
     std::vector<std::complex<double>> energy_r(size);
 
     for (size_t i = 0; i < size; ++i) {
-        energy_k.push_back(wfc_k[i] * h_k[i]);
+        energy_k[i] = wfc_k[i] * h_k[i];
```

Comment:
In split-op this is
```c++
energy_k[i] = wfc_k[i] * pow(complex(par.k[i], 0.0), 2);
```
`h_k` is the same as `par.k`, right?

Reply:
Yes, it should be the same.

Reply:
In that case, the line should read
```suggestion
        energy_k[i] = wfc_k[i] * pow(h_k[i], 2);
```

There is also an extra include for `complex` and an unnecessary include for `cstring` at the top of the file.